text/0088-contract-level-messages.md

```diff
@@ -0,0 +1,237 @@
+
+When the host receives these messages through the `casper_emit_message` FFI, the identity of the caller entity will be attached to the message automatically by the host.
+The messages that are emitted will not be stored in global state; only a checksum of the message will be however stored. This allows users to verify the integrity of the message and also check that the message was indeed generated by the expected entity by doing a query to global state. The entries in global state are uniquely identified by the hash of the entity that emitted the message, the hash of the topic name and the index of the message.
+
+Messages will be passed by the execution engine to the node that will eventually send them out on the event stream after the execution is complete and committed.
+The messages sent out on the event stream contain the identity of the entity that emitted the message, the topic on which the message was emitted, the index of the message within the topic and the actual message payload.
+
+**Note:** the messages themselves cannot be reconstructed by reading the entry from global state since that entry contains only the checksum of the messages sent.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The implementation for supporting contract level messages requires changes in both the execution engine and the node.
+
+### Changes to execution engine
+
+#### Introduce new FFIs:
+```rust
+/// Manages a message topic with the specified name for the calling contract.
+/// 
+/// The topic name is a string of a maximum length that is specified in the chainspec.
+/// The initially supported operation will be `Add` that registers a new message topic.
+/// Will return an error if:
+///     * the topic name length is exceeded
+///     * the topic already exists (in case of the Add operation)
+///     * the call was made from an `Account` context. We only allow stored contracts to register messages.
+pub fn casper_manage_message_topic(
+        topic_name_ptr: *const u8,
+        topic_name_size: usize,
+        operation_ptr: *const u8,
+        operation_size: usize,
+    ) -> i32;
+```
+
+```rust
+/// Emits a new message on the specified topic.
+/// 
+/// The topic needs to have been already registered through a call to `casper_manage_message_topic` before emitting a message.
+/// 
+/// Will return an error if:
+///     * the message is not of a serialized [`MessagePayload`] type.
+///     * the length of the message is larger than the maximum length specified in the chainspec
+pub fn casper_emit_message(
+    topic_name_ptr: *const u8,
+    topic_name_size: usize,
+    message_ptr: *const u8,
+    message_size: usize,
+);
+```
+
+The `MessagePayload` enum initially allows for sending human readable strings as messages in the first iteration:
+```rust
+pub enum MessagePayload {
+    /// Human readable string message.
+    String(String),
```

Comment:
```suggestion
    Bytes(casper_types::bytesrepr::Bytes),
```

Events are blobs of bytes (think CEP18 events). Having them as `String` will force me to to hex or base64 bytes before emitting. It just adds unnecessary gas.

Reply:
Yes, it's probably a good idea to add the `Bytes` variant alongside `String` from the start. The initial requirement was to be able to emit human readable strings, we made the payload an enum in order to allow easy expansion in the future. The enhancement would probably also include standard payloads for system type messages described above.

Reply:
Thanks. Operations on strings in smart contracts are just waste of gas.

Reply:
Added the variant to hold `Bytes` for the payload: https://github.com/casper-network/ceps/pull/88/commits/afaf653819ca6db9bf68415586ffd93aec2dd8a8